choraleutil/protocoltool.1
.TH PROTOCOLTOOL 1 "20-Nov-2008"
.SH NAME
protocoltool \- control an "Empeg Protocol" network device: a car-player or Rio Central

.SH SYNOPSIS

\fBprotocoltool\fP [\fIoptions\fP] [\fIroot-directory\fP [\fIflac-root-directory\fP]]

.SH DESCRIPTION

\fBprotocoltool\fP searches for and controls Empeg (Mark 2)
car-players and Rio Centrals using their inbuilt networking features.

Its main use is to synchronise a directory tree full of media files,
to a car-player or Rio Central, so that the playlist menu structure on
the car-player ends up the same as the directory structure. See the
\fBFILES\fP section for exactly how that works.

.SH OPTIONS

.TP
.B \-d, --device\fP=w.x.y.z
Specify a device (by IP address) rather than assuming there's just one. To
scan for IP addresses of devices, use \fB-s\fP.

.TP
.B \-f, --full-scan
Like \fB-s\fP, but display more information about each discovered device.

.TP
.B \-n, --dry-run
When using the \fB-u\fP operation, don't actually do any synchronising
-- just work out what would need to be transferred.

.TP
.B \-r, --restart
Cause the player to exit and re-start the player firmware in an
orderly fashion.

.TP
.B --reboot
Cause the player to reboot completely via the kernel bootloader.

.TP
.B \-s, --scan
Scan the local network for compatible devices.

.TP
.B \-u, --update
Update (synchronise) the device with the media files in
\fIroot-directory\fP (and perhaps \fIflac-root-directory\fP), or from
the list of media files in the database dump named by \fB--xml\fP. One
of \fIroot-directory\fP and \fB--xml\fP must be given. See the
\fBFILES\fP section for exactly how that works.

.TP
.B \--xml\fP=FILE
When using the \fB-u\fP operation, get the list of media files to
transfer from an XML database file as used by \fBchoraled\fP(1),
rather than directly from the filesystem.

.TP
.B \-h, --help
Display help message and exit.

.SH FILES

If the \fB-u\fP option is given (and the \fB--xml\fP option isn't),
protocoltool will scan the \fIroot-directory\fP, and all directories
inside it, for media files. Symbolic links will be followed, as will
playlist files (*.wpl and *.asx), even if they point to somewhere
outside \fIroot-directory\fP. (Great care is taken to ensure that each
file is transferred only once, even if it's reachable via several
routes.)

Any sub-directories, as well as any playlists, found during this scan
will be represented by nested playlists on the car-player. The
contents of the \fIroot-directory\fP end up as the contents of the
top-level playlist menu on the player.

Note that this is a \fBone-way sync\fP: files which exist on the
player but have no counterpart in \fIroot-directory\fP will be
\fBdeleted from the player\fP.

If you have a car-player with the version 3 firmware (i.e., not
version\ 2, nor a Rio Central, neither of which can play FLAC), you
may want to add a \fIflac-root-directory\fP. This directory will be
scanned in parallel with the \fIroot-directory\fP, and is assumed to
have the same layout. Any time a file is found in
\fIroot-directory\fP, and a corresponding file is found in
\fIflac-root-directory\fP, the latter will be transferred
instead. This eases the business of maintaining parallel collections
of MP3 and FLAC files, as you can make a playlist inside
\fIroot-directory\fP naming a bunch of MP3 files, and still get the
corresponding FLAC files transferred to your car-player. See
\fBchoralecd\fP(1) for a program which will rip CDs into this sort of
parallel directory structure.

.SH RESTRICTIONS

\fBprotocoltool\fP can only access players on Ethernet, not USB or
serial.

There is currently no way to set playlist flags (such as "Ignore as child").

\fBprotocoltool\fP currently assumes your player has enough capacity
to store all your music, and "disk full" errors take it completely by
surprise. Quite frankly, 120GB 2.5in winchesters are forty quid each these
days. Upgrade already!

.SH EXAMPLES

 $ \fBprotocoltool -s\fP
 192.168.1.99 Ernest the Empeg

 $ \fBprotocoltool -d 192.168.1.99 -u -n /home/sarah/mp3 /home/sarah/flac\fP
 \fI...lots of information about what would be transferred...\fP

 $ \fBprotocoltool -d 192.168.1.99 -u /home/sarah/mp3 /home/sarah/flac\fP
 \fI...lots of progress information as the transfers happen...\fP

.SH SEE ALSO

\fBchoraled\fP(1), \fBchoralecd\fP(1)

\fBhttp://chorale.sf.net\fP

\fBhttp://empegbbs.com\fP

.SH AUTHOR

\fBprotocoltool\fP is part of the Chorale collection of multimedia software.

Written by Peter Hartley <pdh@utter.chaos.org.uk>


ChangeLog
9999-XXX-99: Version 0.20 released; changes since 0.19:

	* libdblocal: fix paper-bag bug in file scanning
	* configure: find Cairo includes properly
	* configure: find Boost libs on systems where they have "-mt" suffix
	
2010-Mar-28: Version 0.19 released; changes since 0.18:

	* choralecd: new tageditor
	* choralecd: make user-defined events more sane
	* libdav: RFC4918 WebDAV server
	* libdbsteam: allow setting of field flags in constructor
	* libdblocal: reflect database writes back into the ID3 tags
	* libdblocal: replace FileScannerThread with Task-only DatabaseUpdater
	* libdblocal: add non-waiting API to FileScanner
	* libimport: make sure ID3 rewrite removes old tags
	* libutil: nicer util::Bind syntax
	* libutil: CountedPointer::operator== should be const!
	* libutil: add non-waiting API to DirectoryWalker
	* Make.common: moc files are potentially different per-target
	* configure: deal with <linux/cdrom.h> not being compatible with C++
	* configure: update libtool (needed for Obj-C++)
	152 files changed, 12210 insertions, 6757 deletions (mostly libtool)
	151 files changed, 4863 insertions, 866 deletions (excluding libtool)
	
2009-Oct-18: Version 0.18 released; changes since 0.17:

	* libutil: paper-bag bug in http::Client::Task
	3 files changed, 25 insertions, 9 deletions
	
2009-Oct-18: Version 0.17 released; changes since 0.16:

	* libutil: paper-bag bug in util::Compare (forced pull of 0.16)
	* libutil: Scheduler, like Poller but owns the tasks it waits on
	* libutil: use native threads, not pthreads, on Win32
	* libutil: make HTTP server use Stream (not BufferSink), cleans up code
	* libupnp: asynchronous API
	* libdb: const-correctness for Recordset
	* libmediadb: more forgiving DIDL parsing
	* libtv: record DVB as MPEG PS (not MPEG TS)
	* choralecd: reduce reliance on Qt3Support (not quite eliminated)
	* choralecd: use asynchronous API for initial UPnP output connect
	366 files changed, 14714 insertions, 5229 deletions (since 0.15)
	
2009-Jul-19: Version 0.16 [not, in fact] released; changes since 0.15:

	* choralecd: compile on Win32
	* choralecd: better set-list window with drag-and-drop reorder
	* libdblocal: internet radio support
	* libtv: cope with audio PES stream IDs other than 0xC0
	* libimport: PLS playlist support
	* liboutput: use GstPlayBin2 if available (gapless, allegedly)
	* liboutput: bipartite shuffle support
	* libupnpd: seek support in AVTransport
	* libutil: better support for Unicode filenames on Win32
	* libutil: stream caching for http::Server
	* libutil: case-insensitive sorting of filenames
	* libutil: better Canonicalise for Win32 pathnames
	* libutil: fixes for glibc 2.10.1
	* libutil: wrap boost/mutex.hpp, see http://pdh11.blogspot.com
	* libutil: use homemade counted pointer, not boost::intrusive_ptr
	* libutil: wrappers for ISO file API to pretend Win32 is UTF-8
	* libutil: move WSAAsyncSelect guff from Socket to Poller
	* libutil: util::Compare for caseless accentless comparisons
	* scripts/topten: name and shame the least-test-suite-covered files
	* Make.common: don't gather GCOV statistics from Win32 (doesn't work)
	* Make.common: fixes for gold
	* Make.common: compile with -fno-exceptions where possible
	* configure: deal with --with-[l]gpl=yes correctly
	296 files changed, 9264 insertions, 2658 deletions

2009-May-01: Version 0.15 released; changes since 0.14:

	* choralecd: separate UI code from ripping-control code
	* choraled: unified code Win32/Linux, many new features in Win32
	* choraled: --assimilate-receiver
	* choraled: understand more formats in MediaRenderer
	* choraled: deal with non-ASCII install and music paths in choralesvc
	* choraled: now works out itself what licence it's under and tells you
	* choraleutil, libutil: compilation fixes for GCC 4.4.0
	* protocoltool: understand Jupiters better
	* libimport: slow down ripping if errors are occurring
	* libdbreceiver: fixes for "traditional" receiver servers eg jupiter
	* libmediadb: proxy HTTP access, don't point clients at origin server
	* libdbsteam: query-lifetime fixes (seen as receiver service crashes)
	* libupnp: hopefully-final fix for multihomed hosts
	* libupnp: saner client API that deals with reappearing on changed port
	* libupnp: pay more attention to content-types (text/xml)
	* libutil: SEQUENTIAL flag for file writes
	* libutil: logging API that's driveable by environment variables
	* libutil: ignore hidden files on Windows
	* libutil: filename manipulation now understands backslash on Windows
	* libutil: canonicalise Windows filenames to forward slash
	* libutil: select for FD_CLOSE'd sockets on Windows
	* libutil: http::Server now understands HEAD requests
	* libutil: Scanf64 to work around Mingw scanf/printf %llu severe fail
	* Makefile: allow "make headercheck" for header file quality check
	* Makefile: install XML files on Win32
	* configure: make all HAVE_xyz tests "#if" rather than "#ifdef"
	195 files changed, 4946 insertions, 2035 deletions
	
2009-Mar-05: Version 0.14 released; changes since 0.13:

	* configure, COPYING: new Qt versions are LGPL
	* libdblocal: new home for mediadb::Database
	* libmediadb: more robustness against corrupt databases
	* libdbmerge: new merged view of databases
	* libtv: recordings now asynchronous rather than tying up a thread each
	* libupnp: upgrade SCPD versions from upnp.org
	* libupnp: support service versioning
	* libupnp: ClientConnection::Init now takes service_id not service_type
	* libupnp: ssdp::Responder::Advertise now takes UDN not USN
	* libupnp: SSDP now deals better with multihomed hosts
	* libreceiver: use subnet-directed broadcasts for pseudo-SSDP
	* libempeg: use subnet-directed broadcasts for discovery
	* libutil: clean up error-prone semantics of http::Client::Connect
	* libutil: use libwrap to allow access control by IP address
	* choraled: use libdbmerge to show radio alongside files
	* choraled: only default to -p 12078 if serving Rio Receivers
	127 files changed, 1703 insertions, 756 deletions
	

choraleutil/protocoltool.cpp
#include "config.h"
#include "libdbempeg/db.h"
#include "libmediadb/xml.h"
#include "libmediadb/sync.h"
#include "libmediadb/schema.h"
#include "libdblocal/db.h"
#include "libdbsteam/db.h"
#include "libdblocal/file_scanner.h"
#include "libdb/query.h"
#include "libdb/recordset.h"
#include "libempeg/discovery.h"
#include "libempeg/protocol_client.h"
#include "libutil/scheduler.h"
#include "libutil/socket.h"
#include "libutil/file_stream.h"
#include "libutil/http_server.h"
#include "libutil/http_client.h"
#include "libutil/worker_thread_pool.h"
#include <getopt.h>
#include <stdlib.h>
#if HAVE_SYS_TIME_H
#include <sys/time.h>
#endif

/** Command-line program for Empeg protocol operations (eg for car-players).
 */
namespace protocoltool {

static bool dry_run = false;
static bool do_scan = false;
static bool do_full_scan = false;
static bool do_restart = false;
static bool do_reboot = false;
static bool do_update = false;
static const char *update_root = NULL;
static const char *update_flac_root = NULL;
static const char *update_xml = NULL;
static util::IPAddress s_device = util::IPAddress::ANY;

static void Usage(FILE *f)
{
    fprintf(f,
	 "Usage: protocoltool [options] [<root-directory> [<flac-root-directory>]]\n"
"Control an \"Empeg Protocol\" network device: a car-player or Rio Central.\n"
"\n"
"The options are:\n"
" -d, --device=IP    Select device (rather than picking one at random)\n"
" -s, --scan         Scan network for compatible devices\n"
" -f, --full-scan      Display full information about each device\n"
" -r, --restart      Orderly exit and re-start of player firmware\n"
"     --reboot       Completely reboot kernel\n"
" -u, --update       Update (synchronise) device to match local database\n"
" -n, --dry-run        Don't actually update the device, just print details\n"
"     --xml=FILE       Use Chorale database dump FILE instead of scanning files\n"
" -h, --help         These hastily-scratched notes\n"
	    "\n"

"  Supplying -d is required for -u. Supplying <root-directory> is also\n"
"required for -u, unless --xml is used. The <flac-root-directory> is assumed\n"
"to have the same layout as <root-directory> but with FLACs instead of MP3s;\n"
"if not given, no FLACs are transferred, which is what you want with v2\n"
"car-players as they can't play them anyway."

"\n"
"From " PACKAGE_STRING " (" PACKAGE_WEBSITE ") built on " __DATE__ ".\n"
	);
}

static void Version()
{
    printf("protocoltool, part of " PACKAGE_STRING ".\n");
}


        /* FindDevice */


/** Callback that just picks the first device to respond.
 */
struct FindDeviceCallback: public empeg::Discovery::Callback
{
    void OnDiscoveredEmpeg(const util::IPAddress& ip, const std::string& name)
    {
	s_device = ip;
	printf("Using %s on %s\n", name.c_str(), ip.ToString().c_str());
    }
};

/** Assuming there's only one device on the network, find it and store its
 * IP address in s_device.
 */
void FindDevice()
{
    util::BackgroundScheduler poller;
    empeg::Discovery disc;
    FindDeviceCallback sc;
    disc.Init(&poller, &sc);

    time_t t = time(NULL) + 5;

    while (time(NULL) < t && s_device != util::IPAddress::ANY)
	poller.Poll(1000);

    if (s_device == util::IPAddress::ANY)
    {
	fprintf(stderr, "No devices found\n");
	exit(1);
    }
}


        /* Scan */


/** Callback that lists all responding devices.
 */
struct ScanCallback: public empeg::Discovery::Callback
{
    bool any;

    ScanCallback() : any(false) {}

    void OnDiscoveredEmpeg(const util::IPAddress& ip, const std::string& name);
};

void ScanCallback::OnDiscoveredEmpeg(const util::IPAddress& ip, 
				     const std::string& name)
{
    printf("%-15s %s\n", ip.ToString().c_str(), name.c_str());
    if (do_full_scan)
    {
	empeg::ProtocolClient pc;
	unsigned int rc = pc.Init(ip);
	if (rc != 0)
	{
	    printf("%-15s ** Can't contact: error %u\n", "", rc);
	    return;
	}
	std::string version, id, type, configini;
	rc = pc.ReadFidToString(empeg::FID_VERSION, &version);
	if (!rc)
	    rc = pc.ReadFidToString(empeg::FID_ID, &id);
//	if (!rc)
//	    rc = pc.ReadFidToString(empeg::FID_CONFIGINI, &configini);
	if (!rc)
	    rc = pc.ReadFidToString(empeg::FID_TYPE, &type);
	if (rc)
	{
	    printf("%-15s ** Can't collect info: error %u\n", "", rc);
	    return;
	}
	version.erase(version.find_last_not_of(" \t\n")+1);

	const char *idstr = strstr(id.c_str(), "hwrev :");
	unsigned int hwrev = 0;
	if (idstr)
	    hwrev = atoi(idstr + 7);
	idstr = strstr(id.c_str(), "serial:");
	unsigned int serial = 0;
	if (idstr)
	    serial = atoi(idstr + 7);

	if (any)
	    printf("\n");
	any = true;

	if (!strcmp(type.c_str(), "jupiter"))
	{
	    // Holy hell, it's a Rio Central/HSX-109. Not many of them around.

	    printf("%-15s Jupiter #%06u v%s hwrev %u\n",
		   "",
		   serial % 1000000,
		   version.c_str(),
		   hwrev);
	}
	else
	{
	    // Mark 2 serial numbers are decimal mmyynnnnn
	    //   mm = month of manufacture 01-12
	    //   yy = year of manufacture 00-01
	    //   nnnnn = actual serial number starting at 1
	    //
	    // This means that the very first Mark 2 was #060000001, and
	    // also that serial numbers order isn't chronological order:
	    // 120001100 was manufactured before 010101200 despite being
	    // numerically larger.
	    
	    printf("%-15s Mark %-2s #%05u (%02u/%u) v%s\n",
		   "",
		   (hwrev < 9) ? "2" : "2a", // No Mark 1s on ethernet!
		   serial   % 100000,        // We only planned 100,000 ever!
		   (serial  / 10000000),             // Month of manufacture
		   ((serial / 100000) % 100) + 2000, // Year of manufacture
		   version.c_str());
	}

//	printf("version: '%s'\n", version.c_str());
//	printf("type: '%s'\n", type.c_str());
//	printf("id: >>>%s\n<<<\n", id.c_str());
//	printf("config: >>>%s\n<<<\n", configini.c_str());
    }
    else
	any = true;
}

static int Scan()
{
    util::BackgroundScheduler poller;
    empeg::Discovery disc;
    ScanCallback sc;
    disc.Init(&poller, &sc);

    time_t t = time(NULL);

    while ((time(NULL) - t) < 5)
    {
	poller.Poll(1000);
    }

    if (!sc.any)
    {
	printf("No devices found\n");
	return 1;
    }

    return 0;
}


        /* Update */


static void PrintSeconds(unsigned long long sec)
{
    if (sec == 0)
	sec = 1;

    if (sec > 3600*24)
    {
	printf("%llud", sec/(3600*24));
	sec %= (3600*24);
    }
    if (sec > 3600)
    {
	printf("%lluh", sec/3600);
	sec %= 3600;
    }
    if (sec > 60)
    {
	printf("%llum", sec/60);
	sec %= 60;
    }
    if (sec)
	printf("%llus", sec);
}

class UpdateObserver: public mediadb::SyncObserver
{
    db::Database *m_src;
    db::Database *m_dest;

    void Print(db::Database *db, const char *verb, unsigned int id,
	       size_t num, size_t denom)
    {
	db::QueryPtr qp = db->CreateQuery();
	qp->Where(qp->Restrict(mediadb::ID, db::EQ, id));
	db::RecordsetPtr rs = qp->Execute();
	if (rs && !rs->IsEOF())
	{
	    printf("%s %u/%u %s\n", verb,
		   (unsigned int)num+1,
		   (unsigned int)denom, 
		   rs->GetString(mediadb::TITLE).c_str());
	}
    }

public:
    UpdateObserver(db::Database *src, db::Database *dest)
	: m_src(src), m_dest(dest) {}
	
    void OnAddFile(unsigned int srcid, size_t num, size_t denom)
    {
	Print(m_src, "Adding", srcid, num, denom);
    }

    void OnAddPlaylist(unsigned int srcid, size_t num, size_t denom)
    {
	Print(m_src, "Adding playlist", srcid, num, denom);
    }

    void OnAmendPlaylist(unsigned int srcid, size_t num, size_t denom)
    {
	Print(m_src, "Updating playlist", srcid, num, denom);
    }

    void OnAmendMetadata(unsigned int srcid, size_t num, size_t denom)
    {
	Print(m_src, "Updating", srcid, num, denom);
    }
};

static int Update()
{
    util::WorkerThreadPool wtp(util::WorkerThreadPool::NORMAL, 32);

    // Prepare source database
    db::steam::Database sdbsrc(mediadb::FIELD_COUNT);
    sdbsrc.SetFieldInfo(mediadb::ID, 
		     db::steam::FIELD_INT|db::steam::FIELD_INDEXED);
    sdbsrc.SetFieldInfo(mediadb::PATH,
		     db::steam::FIELD_STRING|db::steam::FIELD_INDEXED);
    sdbsrc.SetFieldInfo(mediadb::TITLE,
		     db::steam::FIELD_STRING|db::steam::FIELD_INDEXED);
    sdbsrc.SetFieldInfo(mediadb::TRACKNUMBER,
		     db::steam::FIELD_INT|db::steam::FIELD_INDEXED);

    util::http::Client http_client;
    db::local::Database dbsrc(&sdbsrc, &http_client);

    unsigned int rc;

    if (update_xml)
    {
	std::auto_ptr<util::Stream> fsp;
	rc = util::OpenFileStream(update_xml, util::READ, &fsp);
	if (rc)
	{
	    fprintf(stderr, "Can't open %s: %u\n", update_xml, rc);
	    return 1;
	}
	rc = mediadb::ReadXML(&dbsrc, fsp.get());
	if (rc)
	{
	    fprintf(stderr, "Can't parse %s: %u\n", update_xml, rc);
	    return 1;
	}
    }
    else
    {
	db::local::FileScanner ifs(update_root, update_flac_root, &sdbsrc,
				   &dbsrc, &wtp);

	rc = ifs.Scan();
	if (rc)
	{
	    fprintf(stderr, "Can't scan %s: %u\n", update_root, rc);
	    return 1;
	}
    }

    // Prepare destination database
    util::BackgroundScheduler poller;
    util::http::Server server(&poller, &wtp);
    db::empeg::Database dbdest(&server);
    rc = dbdest.Init(s_device);
    if (rc != 0)
    {
	fprintf(stderr, "Can't contact %s: error %u\n", 
		s_device.ToString().c_str(), rc);
	return 1;
    }

    struct timeval tv;
    ::gettimeofday(&tv, NULL);
    uint64_t usec = (((uint64_t)tv.tv_sec) * 1000000) + tv.tv_usec;

    // Perform sync
    mediadb::Synchroniser sync(&dbsrc, &dbdest, dry_run);
    UpdateObserver uobs(&dbsrc, &dbdest);
    sync.AddObserver(&uobs);

    rc = sync.Synchronise();
    if (rc != 0)
    {
	fprintf(stderr, "Synchronise failed: error %u\n", rc);
	return 1;
    }

    ::gettimeofday(&tv, NULL);
    usec = (((uint64_t)tv.tv_sec) * 1000000) + tv.tv_usec - usec;

    mediadb::Synchroniser::Statistics s = sync.GetStatistics();
    if (dry_run)
    {
	printf("Would transfer %u files, %u playlists, %llu bytes;\n"
	       "  delete %u files, %u playlists; amend %u files and %u playlists.\n",
	       s.files_added, s.playlists_added, s.bytes_transferred,
	       s.files_deleted, s.playlists_deleted, s.files_modified,
	       s.playlists_modified);
	unsigned long long estimated_sec = s.bytes_transferred / 750000;
	if (estimated_sec)
	{
	    printf("Estimated sync time ");
	    PrintSeconds(estimated_sec);
	    printf("\n");
	}
    }
    else
    {
	printf("Transferred %u files, %u playlists, %llu bytes;\n"
	       "  deleted %u files, %u playlists; amended %u files and %u playlists.\n",
	       s.files_added, s.playlists_added, s.bytes_transferred,
	       s.files_deleted, s.playlists_deleted, s.files_modified,
	       s.playlists_modified);
	printf("Sync time ");
	PrintSeconds(usec/1000000);
	printf("\n");
    }

    return 0;
}


        /* Restart/reboot */


static int QuitPlayer(unsigned int (empeg::ProtocolClient::*f)(void))
{
    empeg::ProtocolClient pc;
    unsigned int rc = pc.Init(s_device);
    if (!rc)
	rc = (pc.*f)();
    if (rc)
    {
	fprintf(stderr, "Can't contact: error %u\n", rc);
	return 1;
    }
    printf("OK\n");
    return 0;
}


        /* Main program */


int Main(int argc, char *argv[])
{
    static const struct option options[] =
    {
	{ "help", no_argument, NULL, 'h' },
	{ "version", no_argument, NULL, 'v' },
	{ "device", required_argument, NULL, 'd' },
	{ "update", no_argument, NULL, 'u' },
	{ "dry-run", no_argument, NULL, 'n' },
	{ "scan", no_argument, NULL, 's' },
	{ "full-scan", no_argument, NULL, 'f' },
	{ "restart", no_argument, NULL, 'r' },
	{ "reboot", no_argument, NULL, 1 },
	{ "xml", required_argument, NULL, 2 }
    };

    int option_index;
    int option;
    while ((option = getopt_long(argc, argv, "hvsfnd:ur", options,
				 &option_index))
	   != -1)
    {
	switch (option)
	{
	case 'h':
	    Usage(stdout);
	    return 0;
	case 'v':
	    Version();
	    return 0;
	case 'd':
	    s_device = util::IPAddress::Resolve(optarg);
	    break;
	case 'n':
	    dry_run = true;
	    break;
	case 'u':
	    do_update = true;
	    break;
	case 's':
	    do_scan = true;
	    break;
	case 'f':
	    do_scan = true;
	    do_full_scan = true;
	    break;
	case 'r':
	    do_restart = true;
	    break;
	case 1:
	    do_reboot = true;
	    break;
	case 2:
	    update_xml = optarg;
	    break;
	default:
	    Usage(stderr);
	    return 1;
	}
    }

    if (do_update && !update_xml)
    {
	if (argc > optind+2 || argc <= optind)
	{
	    Usage(stderr);
	    return 1;
	}
	update_root      = argv[optind];
	update_flac_root = argv[optind+1];
    }
    else
    {
	if (argc != optind)
	{
	    Usage(stderr);
	    return 1;
	}
    }

    if (!do_scan && !do_update && !do_restart && !do_reboot)
    {
	Usage(stdout);
	return 0;
    }

    if (do_scan)
	return Scan();

    if (do_restart)
    {
	if (s_device == util::IPAddress::ANY)
	    FindDevice();
	return QuitPlayer(&empeg::ProtocolClient::RestartPlayer);
    }

    if (do_reboot)
    {
	if (s_device == util::IPAddress::ANY)
	    FindDevice();
//	return QuitPlayer(&empeg::ProtocolClient::Reboot);
    }

    if (do_update)
    {
	if (s_device == util::IPAddress::ANY)
	{
	    fprintf(stderr, "Specifying --device is necessary for --update. (Hint: --scan.)\n");
	    return 1;
	}
	if (!update_xml && !update_root)
	{
	    fprintf(stderr, "Specifying either --xml or a root directory is necessary for --update.\n");
	    return 1;
	}
	return Update();
    }
    
    return 0;
}

} // namespace protocoltool

int main(int argc, char *argv[])
{
    return protocoltool::Main(argc, argv);
}
